packages/graphql-plugin-auth/src/JWKSPlugin.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import jsonwebtoken from "jsonwebtoken";
+import JwksRsa, { JwksClient } from "jwks-rsa";
+import Debug from "debug";
+import { DEBUG_PREFIX } from "./constants";
+
+const debug = Debug(DEBUG_PREFIX);
+
+function verifyJWKS<T = any>({ client, token }: { client: JwksClient; token: string }): Promise<T> {
+    const getKey: jsonwebtoken.GetPublicKeyOrSecret = (header, callback) => {
+        const kid: string = header.kid || "";
+
+        client.getSigningKey(kid, (err, key) => {
+            const signingKey = key?.getPublicKey();
+            callback(err, signingKey);
+        });
+    };
+
+    // Returns a Promise with verification result or error
+    return new Promise((resolve, reject) => {
+        jsonwebtoken.verify(
+            token,
+            getKey,
+            {
+                algorithms: ["HS256", "RS256"],
+            },
+            (err, decoded) => {
+                if (err) {
+                    reject(err);
+                } else {
+                    resolve(decoded as unknown as T);
+                }
+            }
+        );
+    });
+}
+
+export interface JWKSPluginInput {
+    jwksEndpoint: string;
+    rolesPath?: string;
+    clientOptions?: JwksRsa.Options;
+}
+
+class JWKSPlugin {
+    jwksEndpoint: string;
+    rolesPath?: string;
+    client: JwksClient;
+
+    constructor(input: JWKSPluginInput) {
+        this.jwksEndpoint = input.jwksEndpoint;
+        this.rolesPath = input.rolesPath;
+        let options: JwksRsa.Options | undefined = input.clientOptions;
+
+        if (!options) {
+            const defaultOptions: JwksRsa.Options = {
+                jwksUri: this.jwksEndpoint,
+                rateLimit: true,
+                jwksRequestsPerMinute: 10,
+                cache: true,
+                cacheMaxEntries: 5,
+                cacheMaxAge: 600000,
+            };
+
+            options = defaultOptions;
+        }
```

Comment:
**question/suggestion (blocking)**: What happens if the user only defines one of the options?

Consider the following example:
```typescript
new JWKSPlugin({
    jwksEndpoint: "cerveza",
    rolesPath: "beer",
    clientOptions: {
         jwksUri: "not-cerveza"
    }
})
```

In this example,  we have two issues:
* The clients `jwksUri` does not match the class `jwksEndpoint`
* All unset defaults are ignored (e.g. `rateLimit`)

This feels like it is wrong, if this is the intended way of working, please ignore the rest of this comment

Maybe we could consider merging both `options` and `defaultOptions` to ensure all the defaults are set and that jwksEndpoint is valid. 

This is a possible way of doing it, taking care of the typings:
```suggestion
        const defaultOptions: Omit<JwksRsa.Options, "jwksUri> = {
              rateLimit: true,
              jwksRequestsPerMinute: 10,
              cache: true,
              cacheMaxEntries: 5,
              cacheMaxAge: 600000,
        };
            
       const clientOptions: Partial<JwksRsa.Options> = input.clientOptions || {};
       const options: JwksRsa.Options = {{jwksUri: input.jwksEndpoint},...defaultOptions, ...input.clientOptions };
       this.jwksEndpoint=options.jwksUri; // This way, jwksUri and jwksEndpoint are always the same
```

Also, fiddling with this, we may want to change `JWKSPluginInput` to use `Partial<JwksRsa.Options>` for `clientOptions`

Reply:
I've removed this for now because it was scope creep and it's not tested, good observation.